031-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,156 @@
+# TLS encrypting the Kafka Connect REST API
+
+This proposal adds the option to reduce access to the Kafka Connect REST API by enabling TLS encryption on the Kafka Connect REST listener.
+
+## Current situation
+
+Currently, instances of Kafka Connect that are deployed by the Strimzi operators are configured with the default REST API endpoint settings.
+This means that the Kafka Connect REST API endpoint uses HTTP on port 8083 and that the Strimzi `KafkaConnect`, `KafkaConnector` and `KafkaMirrorMaker2` operators make unsecured REST client calls based on this default configuration.
+
+The default network policies created by the Strimzi operators restrict incoming REST API calls to only allow access from the operator pod.
+Users can define further network policies to override the default policy and allow wider access.
+
+## Motivation
+
+There is currently no way to TLS encrypt the Kafka Connect REST API, leaving the Connect REST listener as perhaps the last key endpoint that cannot currently be restricted using Strimzi.
+
+## Proposal
+
+This proposal adds REST api configuration options to the `KafkaMirrorMaker2` and `KafkaConnect` CRDs.
+
+A new field is added to the `KafkaConnect` and `KafkaMirrorMaker2` specs named `restListeners`, which allows users to configure an HTTP listener on port 8443, an encrypted HTTPS listener on port 8443, or both in the Kafka Connect configuration.
+This enables users to access the REST API with TLS if required.
+The `KafkaConnector`, `KafkaConnect` and `KafkaMirrorMaker2` operators will use the TLS encrypted listener when it is enabled, even if an unencrypted listener is present.
+
+The `restListeners` field is designed for potential future extension, and is a list of items that contains a `protocol` field and an optional `restCertChainAndKey` field.
+
+For this proposal, the `protocol` field only supports two values: `http` and `https`. If the `restListeners` list is empty the `KafkaConnect` and `KafkaMirrorMaker2` 
+runtimes will be created with a single unencrypted listener on 8083 to match the existing behaviour.
+
+The `restCertChainAndKey` field will be required if the `https` protocol is selected. This field will contain the certificate and key the Kafka Connect operator will 
+use to create the self-signed certificate. The Kafka Connect and MirrorMaker2 operators will use the certificate to create a truststore for communicating with the REST API.
```

Comment:
The above two sentences are not pretty clear to me.
So, to enable HTTPS on the REST API we need to generate a public cert and a private key (and it should be the KafkaConnect operator going to do so); the public cert will be the one provided by the HTTPS server to the client during the TLS handshake and it has to be signed by a "CA" certificate which will be "imported" in the truststore of the clients to trust the public cert got from the HTTPS server.
If the HTTPS server is providing a self-signed certificate to the client, it means that the client has to have exactly that certificate in its truststore to accept the TLS handshake but I don't think it's the way we want, right?
I don't think this process is really clear from the above description. 

> ...  to create the self-signed certificate

What's the self-signed certificate you are mentioning there?

> The Kafka Connect and MirrorMaker2 operators will use the certificate to create a truststore ....

Which one are you referring to?

Reply:
@ppatierno I've updated the document and added a specific section the discusses the certificates, let me know if it is clearer now.